### PR TITLE
impl Num.neg for Dec,F32,F64 in repl

### LINKED ISSUE #6959: `Num.neg` not supported for `Dec`, or for any `Num` except I64/U64 when in repl
When trying to build the following test program:
```
app [main] {
    cli: platform "https://github.com/roc-lang/basic-cli/releases/download/0.12.0/Lb8EgiejTUzbggO2HVVuPJFkwvvsfW6LojkLR20kTVE.tar.br",
}
import cli.Stdout

a : Dec
a = 10

main =
    Num.neg a
    |> Num.toStr
    |> Stdout.line
```
I get the output:
```
thread 'main' panicked at crates/compiler/gen_llvm/src/llvm/lowlevel.rs:2230:13:
internal error: entered unreachable code: Unrecognized dec unary operation: NumNeg
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

When investigating the source, it's clear that `Num.neg` is simply not implemented for `Dec`.
This isn't mentionend in the documentation or anywhere else I've looked.

Additionally, using `Num.neg` on any `Num` that's not `I64` or `U64` is not supported from the repl. For example:
```
» a : I32
… a = 10

10 : I32
» -a
thread 'main' panicked at crates/compiler/gen_dev/src/generic64/mod.rs:1794:18:
not yet implemented: NumNeg: layout, Builtin(Int(I32))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
You can of course work around this by subtracting from 0, but I think this should at least be documented.

### PR DESCRIPTION
**Pull request by shua**
The intent of this change is to support `Num.neg` for Dec, F32, and F64 types in the repl.

Dec negation was implemented across gen-dev, gen-llvm, gen-wasm as a call to the compiled zig function `bitcode::DEC_NEGATE`.

f32 and f64 negation were implemented already for gen-llvm, gen-wasm.

for gen-dev x86_64, float negation is implemented by flipping the sign bit, which means `xorps` for f32, and `xorpd` for f64

for gen-dev aarch64, there is conveniently a `fneg` instruction

Closes #6959

### CODE REVIEW

File: crates/compiler/gen_dev/src/generic64/mod.rs [hunk 1]
```diff
@@ -1791,7 +1791,38 @@ impl<
...
+                let src_reg = self.storage_manager.load_to_float_reg(&mut self.buf, src);
+                self.storage_manager
+                    .with_tmp_float_reg(&mut self.buf, |_, buf, tmp| {
+                        // gcc and llvm will do `f ^ 0x8000_0000` to flip the sign bit
+                        // but we are doing 0-f here because I couldn't figure out how
+                        // to load floats into general registers
+                        ASM::mov_freg32_imm32(buf, &mut self.relocs, tmp, 0.0f32);
+                        ASM::sub_freg32_freg32_freg32(buf, dst_reg, tmp, src_reg);
+                    })
+            }
```
Comment (lukewilliamboswell) [inline crates/compiler/gen_dev/src/generic64/mod.rs:1805]: @folkertdev 

Do you know how to load floats into general registers?

Comment (shua) [reply crates/compiler/gen_dev/src/generic64/mod.rs:1805]: I may have been mistaken about that. The equivalent code in C is compiled to use `xmm` registers which are floating point registers. There's just no `xorps` or `xorpd` ops defined yet for gen-llvm (or equivalent "bitwise xor of floating point register" for gen-wasm/gen-dev), but I can work on that.

[godbolt C example](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:2,endLineNumber:7,positionColumn:2,positionLineNumber:7,selectionStartColumn:2,selectionStartLineNumber:7,startColumn:2,startLineNumber:7),source:'/*+Type+your+code+here,+or+load+an+example.+*/%0Adouble+neg_f64(double+f)+%7B%0A++++return+-f%3B%0A%7D%0Afloat+neg_f32(float+f)+%7B%0A++++return+-f%3B%0A%7D'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:39.03462749213011,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cg142,filters:(b:'0',binary:'1',binaryObject:'0',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'',overrides:!(),selection:(endColumn:30,endLineNumber:7,positionColumn:30,positionLineNumber:7,selectionStartColumn:30,selectionStartLineNumber:7,startColumn:30,startLineNumber:7),source:1),l:'5',n:'0',o:'+x86-64+gcc+14.2+(Editor+%231)',t:'0')),k:60.96537250786989,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

Comment (lukewilliamboswell) [reply crates/compiler/gen_dev/src/generic64/mod.rs:1805]: Worth giving it a shot 😄

**Comment by lukewilliamboswell** [comment]
This is looking good @shua 😍 

That CI failure looks genuine
```
thread 'gen_num::num_negate' panicked at crates/compiler/gen_wasm/src/low_level.rs:1627:26:
not yet implemented: NumNeg for InLayout(DEC)
```

**Comment by shua** [comment]
ah yeah, I was running `cargo test -p test_gen num_negate` and wasn't hitting the error, but I guess it's a different test. Looks like a simple addition to `gen_wasm`.

File: crates/compiler/gen_dev/src/generic64/mod.rs [hunk 2]
```diff
@@ -557,6 +557,18 @@ pub trait Assembler<GeneralReg: RegTrait, FloatReg: RegTrait>: Sized + Copy {
...
+        relocs: &mut Vec<'_, Relocation>,
+        dst: FloatReg,
+        src: FloatReg,
+    );
+    fn neg_freg32_freg32(
+        buf: &mut Vec<'_, u8>,
+        relocs: &mut Vec<'_, Relocation>,
+        dst: FloatReg,
+        src: FloatReg,
+    );
```
Comment (shua) [inline crates/compiler/gen_dev/src/generic64/mod.rs:571]: should these just be one `neg_freg_freg` that takes a `FloatWidth` or is it better to have two?

Comment (lukewilliamboswell) [reply crates/compiler/gen_dev/src/generic64/mod.rs:571]: I think two is fine. 👍

**Comment by shua** [comment]
it looks like the CI failure was a test timeout that looks unrelated to my change

> test glue_cli_tests::rust_basic_record has been running for over 60 seconds

**Comment by Anton-4** [comment]
> it looks like the CI failure was a test timeout that seems unrelated to my change

glue_cli_tests failed on the second run too, we'll have to dig in to this.

**Comment by shua** [comment]
yes, I'll try to reproduce locally tonight to see what's going on

**Comment by shua** [comment]
Probably my local machine is beefier than the CI, but after `cargo clean` the test finished in ~13seconds. Not sure why it's hanging in CI. 
```
$ cargo test --release --features=wasm32-cli-run glue_cli_tests --timings

 ---8<--- bunch of irrelevant test output snipped ---8<---

     Running tests/test_glue_cli.rs (target/release/deps/test_glue_cli-9d2fe7f4b1b4108a)

running 18 tests
test glue_cli_tests::all_fixtures_have_tests ... ignored
test glue_cli_tests::c_hello_world ... ok
test glue_cli_tests::rust_option ... ok
test glue_cli_tests::rust_nullable_unwrapped ... ok
test glue_cli_tests::rust_nested_record ... ok
test glue_cli_tests::rust_nonnullable_unwrapped ... ok
test glue_cli_tests::rust_closures ... ok
test glue_cli_tests::rust_nullable_wrapped ... ok
test glue_cli_tests::rust_advanced_recursive_union ... ok
test glue_cli_tests::rust_enumeration ... ok
test glue_cli_tests::rust_rocresult ... ok
test glue_cli_tests::rust_multiple_modules ... ok
test glue_cli_tests::rust_single_tag_union ... ok
test glue_cli_tests::rust_list_recursive_union ... ok
test glue_cli_tests::rust_basic_recursive_union ... ok
test glue_cli_tests::rust_basic_record ... ok
test glue_cli_tests::rust_union_with_padding ... ok
test glue_cli_tests::rust_union_without_padding ... ok

test result: ok. 17 passed; 0 failed; 1 ignored; 0 measured; 0 filtered out; finished in 13.28s
```

I'm about out of ideas for this, I'm not even sure what specifically the `glue_cli_tests` are doing, or why my change would cause one test to hang.

**Comment by lukewilliamboswell** [comment]
I also am having trouble seeing how this change could be related to this test 🤷

**Comment by shua** [comment]
There's a comment in the glue tests about #5924 those tests can cause segfaults in some instances maybe due to misalignment? _Maybe_ changing the gen_wasm crate changes some alignment? But that ticket lists a couple tests which are flakey whereas the issue here seems to *only* be with `glue_cli_tests::rust_nullable_unwrapped`

My other thought is it seems like this test is compiling some rust code as part of the test and maybe some build cache is getting cleared and for some reason it takes more than a minute to recompile from scratch, but that feels even less likely than the flakey alignment issue thing.

:thinking: a real chin-scratcher, really doesn't help that I can't reproduce locally

**Comment by Anton-4** [comment]
I also could not reproduce this on my machine, I'll try to got some more info from the CI machine.

**Comment by Anton-4** [comment]
This could be the first occurrence of #5924 with the legacy linker

**Comment by Anton-4** [comment]
valgrind does not report any problems, so this does not look like https://github.com/roc-lang/roc/issues/5924.

**Comment by Anton-4** [comment]
Alright, the issue does not lie with this PR, I was also able to reproduce it on main. Fixing/investigating this could take very long so I'm going to apply a workaround; run the glue_cli_tests single threaded.

**Comment by shua** [comment]
Not sure if status is "reviewers are busy and aren't constantly checking status of PRs, so please ping when ready" or "reviewers are busy and working through list of PRs to review; please don't ping, we'll get to you".

@lukewilliamboswell should I check back in a week or two, or should I ping somebody else now?

**Comment by Anton-4** [comment]
I got in touch with Brendan, he should be able to review today

**Comment by bhansconnect** [comment]
Yeah, sorry about being slow. I have more or less given up on all of my github emails. And life has been busy so I don't check this nearly as much as I used to.

File: .github/workflows/nix_linux_x86_64.yml [hunk 3]
```diff
@@ -24,7 +24,12 @@ jobs:
         run: nix develop -c ./ci/roc_test_builtins.sh
 
       - name: test wasm32 cli_tests
-        run: nix develop -c cargo test --locked --release --features="wasm32-cli-run"
+        # skipping glue tests due to difficult multithreading bug, we run them single threaded in the next step
+        run: nix develop -c cargo test --locked --release --features="wasm32-cli-run" -- --skip glue_cli_tests
+
+      - name: wasm32 glue_cli_tests
+        # single threaded due to difficult bug when multithreading
+        run: nix develop -c cargo test --locked --release --features="wasm32-cli-run" glue_cli_tests -- --test-threads=1
```
Comment (bhansconnect) [inline .github/workflows/nix_linux_x86_64.yml:32]: This feels like it shouldn't be part of this PR. Nothing touches wasm.

File: crates/compiler/gen_dev/src/generic64/aarch64.rs [hunk 4]
```diff
@@ -3953,6 +3973,24 @@ fn fsub_freg_freg_freg(
     buf.extend(inst.bytes());
 }
 
+/// `FNEG Sd/Dd, Sn/Dn`
+#[inline(always)]
+fn fneg_freg_freg(
```
Comment (bhansconnect) [inline crates/compiler/gen_dev/src/generic64/aarch64.rs:3978]: I'm so sorry that our naming scheme has lead to this... haha

**Review by bhansconnect — APPROVED** [review]
Looks great!

File: .github/workflows/nix_linux_x86_64.yml (hunk 3 above)
Comment (bhansconnect) [reply .github/workflows/nix_linux_x86_64.yml:32]: Ah, I see that @Anton-4  added this as a workaround of some sort.

Comment (shua) [reply .github/workflows/nix_linux_x86_64.yml:32]: I see you merged, but just for posterity: we are all confused why this PR seems to trigger some strange behaviour with this test. @Anton-4 added this to help CI accept the PR.

File: crates/compiler/gen_dev/src/generic64/aarch64.rs (hunk 4 above)
Comment (shua) [reply crates/compiler/gen_dev/src/generic64/aarch64.rs:3978]: in my head, I pronounce `freg` as "eff rej" but `fneg` as "eff neg"

File: .github/workflows/nix_linux_x86_64.yml (hunk 3 above)
Comment (Anton-4) [reply .github/workflows/nix_linux_x86_64.yml:32]: Yeah, not caused by this PR but I just included the fix here

**Pull request merged**